lib/src/Base/Algo/CubaIntegration.cxx

```diff
@@ -0,0 +1,315 @@
+#include "openturns/Exception.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+int computeIntegrand(const int *ndim, const double x[], const int *ncomp, double f[], void *userdata);
+
+/**
+ * @class CubaIntegration
+ */
+
+CLASSNAMEINIT(CubaIntegration)
+
+static const Factory<CubaIntegration> Factory_CubaIntegration;
+
+/* Constructor without parameters */
+CubaIntegration::CubaIntegration()
+  : IntegrationAlgorithmImplementation()
+  , epsRel_(1.E-4)
+  , epsAbs_(1.E-4)
+  , optRoutine_("vegas")
+  , maxeval_(1000000)
+  , flags_(2)
+{
+  // Nothing to do
+}
+
+/* Parameters constructor */
+CubaIntegration::CubaIntegration(const String optRoutine)
+  : IntegrationAlgorithmImplementation()
+  , epsRel_(1.E-4)
+  , epsAbs_(1.E-4)
+  , optRoutine_(optRoutine)
+  , maxeval_(1000000)
+  , flags_(2)
+{
+  // Check the name of the routine
+  if (optRoutine != "cuhre" && optRoutine != "divonne" && optRoutine != "suave" && optRoutine != "vegas") throw InvalidArgumentException(HERE) << "Error: the name of the optimisation routine should be \"cuhre\", \"divonne\", \"suave\", or \"vegas\", here optRoutine=" << optRoutine;
+}
+
+/* Virtual constructor */
+CubaIntegration * CubaIntegration::clone() const
+{
+  return new CubaIntegration(*this);
+}
+
+/*
+ * The friend C routine to compute the integrand, that is provided to the Cuba routine.
+ * It calls the function one wants to integrate after changing coordinates and
+ * multiplying by the volume of the interval as Cuba routines perform integration
+ * over the unit hypercube.
+ */
+int computeIntegrand(const int *ndim, const double x[], const int *ncomp, double f[], void *userdata)
+{
+  Point ptIn(*ndim);
+  void** ppFunctionInterval = (void**)(userdata);
+  const Function* const pFunction = (Function*)(ppFunctionInterval[0]);
+  const Interval* const pInterval = (Interval*)(ppFunctionInterval[1]);
+  const Scalar intervalVolume = pInterval->getVolume();
+  const Point lb = pInterval->getLowerBound();
+  const Point ub = pInterval->getUpperBound();
+  for (UnsignedInteger i = 0 ; i < *ndim ; i++)
+  {
+    ptIn[i] = lb[i] + (ub[i] - lb[i]) * *(x + i);
+  }
+  /* Cuba computes an integral over the unit hypercube, one has to multiply its
+     output by the volume of the true integration interval. */
+  const Point ptOut(intervalVolume * (*pFunction)(ptIn));
+  std::copy(ptOut.begin(), ptOut.end(), f);
+
+  return 0;
+}
+
+/* Compute an approximation of \int_{I}f(x)dx, where I
+ * is an multidimensional interval and f a multidimensional function
+ */
+Point CubaIntegration::integrate(const Function & function,
+                                 const Interval & interval) const
+{
+  if (function.getInputDimension() != interval.getDimension()) throw InvalidArgumentException(HERE) << "Error: interval dimension and input dimension have to match, here interval dimension="<< interval.getDimension() << " and input dimension=" << function.getInputDimension();
```

Comment:
here too the code must be guarded by OPENTURNS_HAVE_CUBA, else throw a NotYetImplementedException exception (you can have a look at NLopt class for example that does something similar)